Core Schema Typescript Library

queerviolet opened the pull request:
Initial work on the `core-schema-js` library. We can now:
- bootstrap a schema — find its `@core(using:core)` directive, which may be renamed, and find all the specs it references.
- extract *specified metadata*, based on its spec url. This handles all the naming logic.
- ask questions like, "is this node in the export schema?" and "what is the parent of this node?"
- derive an export schema from a core schema, stripping away all machinery and respecting `@core(export:)` directives.

I wrote a bit more in the README, which I will paste here too:

# Code Overview
- `__tests__/` has tests (not enough yet)
- `serde/` contains serializers and deserializers for data in the AST. This converts between AST nodes
and arbitrary TypeScript objects (typically scalars and plain objects though). The SerDe module doesn't
know anything about specs or namespaces, it just knows how to extract data from the AST in a particular
shape.
- `bind/` contains the data binding layer. This is where we bind *specified* metadata in the AST
to TypeScript objects. Functions in the `bind` layer take a spec, and the first thing they do when
queried is to look up the name of that spec within the document and use it to filter their metadata
sources (which are all directives at this point, but not forever). You can see an example of
the binding layer in action in the [binding for the core spec](./src/specs/core.ts).
- `schema.ts` contains bootstrapping logic and `fromSource`, which quickly generates a pipeline for processing a `DocumentNode`.
- `data.ts` defines `data` and `derive`. These define additional "columns" on AST Nodes (or really, on any JS object whatsoever). We use them frequently to annotate the AST with additional information, including up-tree links (to the parent and document) and parsed metadata. This just creates a symbol and looks it up on the object, but those implementation details are hidden. You can call `set(object, someData, value)` to set `someData` on `object` to `value`, and `get(object, someData)` or `someData(object)` to get the value on the object. You cannot `set` `derive`d values. They are cached on the object, but are read-only.

# Conventions
- Deserializers are expected to fail with some frequency. `Error`s are somewhat expensive to create. Also, it is frequently desirable to process e.g. *all* items in a list (even after one has failed to deserialize)—this ensures that we discover and surface all document issues, beyond the one failure. Thus, rather than creating `Error`s for every single validation error, deserializers return `Result`s, which are either `Ok<T>` or `Err`. `Err`s are not `Error`s—they are small objects, and do not collect a stack on creation. You can call `.toError()` on an `Err` to create an actual JS error with a stack trace.
- All errors which can occur on the document are declared with [`ERR`](./src/err.ts). This returns an `Err` creator
function.
- Some types are named like this: `type Set_TypeOf<S extends Set> = ...`. This convention indicates that the type is a "type function" which extracts the `TypeOf` a `Set`.

queerviolet commented on src/is.ts:
```diff
@@ -0,0 +1,45 @@
+export const isTemplate = (args: any[]): args is Template =>
+  knownTemplates.has(args) || !!(
+    Array.isArray(args[0]) &&
+    !args[0].find(x => typeof x !== 'string') &&
```
done

queerviolet commented on src/is.ts:
```diff
@@ -0,0 +1,45 @@
+  )
+
+export const isAsString = (args: any[]): args is AsString =>
+    isTemplate(args) || typeof args[0] === 'string'
```
done

queerviolet commented on src/is.ts:
```diff
@@ -0,0 +1,45 @@
+    isTemplate(input)
+      ? String.raw(...input as Template)
+      : input[0]
+  ) as any
```
sure!

queerviolet commented on src/is.ts:
```diff
@@ -0,0 +1,45 @@
+
+import type { ASTNode } from 'graphql'
+
+export function isAst<N extends ASTNode>(obj: any, kind: N["kind"]): obj is N {
```
it is now! (also changed the typings to generalize over kind and infer the node type constraint, which is more usable)

queerviolet commented on src/data.ts:
```diff
@@ -0,0 +1,95 @@
+const getValue = Symbol('Get column value')
+const setValue = Symbol('Set column value')
+
+export interface Set<T, S=any> {
```
Done. Changed `Set` to `Write` and `Get` to `Read`

queerviolet commented on src/data.ts:
```diff
@@ -0,0 +1,95 @@
+  = Get<T, S, Default> & Set<T, S>
+
+export interface SetFn<T, S=any> {
+  (source: S, value: T): void
```
done

queerviolet commented on src/index.ts:
```diff
@@ -0,0 +1,11 @@
+export * from './schema'
+export * from './err'
+export * from './data'
```
They're intended for some but not all consumers to use. Basically, if you have additional data that you'd like to attach or cache on AST nodes, you can use the `data` utilities to provide it. Many users won't; some will.

martijnwalraven replied:
While I appreciate the usefulness of something like `data` and `derive`, I agree with @glasser that we should be careful about exposing unfamiliar patterns as part of our public API. Or even as part of our internal implementation, because this has definitely also made understanding the code challenging for me.

martijnwalraven replied:
I think if we want to commit to these patterns, we need to carefully weigh the pros and cons. Adopting this makes onboarding new developers or getting external contributions harder for example. This commitment makes the most sense to me if theuse isn't limited to this library but also used in other parts of our codebase. I also wonder if there aren't existing alternatives we could use instead. I know @benjamn has his own [reactive caching library](https://github.com/benjamn/optimism) that is used in Apollo Client. And there may be other libraries out there that implement some kind of incremental computation.

glasser replied:
(Going through the comments before looking at updates to the code.) I do think Ben's model of publishing separate (not necessarily super documented or explicitly supported) packages of support code that can be consumed by AC without needing to be part of its API is a choice that works out well (though I have slight concerns if this means that changes to the Apollo package that require improvements to the support code need to be gated through one individual)

queerviolet commented on src/is.ts:
```diff
@@ -0,0 +1,45 @@
+    isTemplate(args) || typeof args[0] === 'string'
+
+export type AsTemplateOr<T> = Template | [T]
+export type AsString = AsTemplateOr<string>
```
done, removed the template strings, which you're right, are not adding value.

queerviolet commented on src/data.ts:
```diff
@@ -0,0 +1,95 @@
+  return col[getValue](source, defaultValue)
+}
+
+export function withGet<B, Col extends Get<any, any>>(base: B, col: Col): B & Col {
```
removed

queerviolet commented on package.json:
```diff
@@ -0,0 +1,46 @@
+    "access": "public"
+  },
+  "dependencies": {
+    "apollo-graphql": "^0.6.0",
```
oops, those were in a package.json i copied. removed.

queerviolet commented on src/spec.ts:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  toString() {
+    return `${this.identity}/${this.version}`
```
Added a `url` getter and made `toString()` return more of an informational string

queerviolet commented on src/version.ts:
````diff
@@ -0,0 +1,140 @@
+   * # Example
+   * ```
+   * use using::Version;
+   * assert_eq!(Version::parse("v1.0")?, Version(1, 0));
````
ah yes. done

queerviolet commented on src/version.ts:
```diff
@@ -0,0 +1,140 @@
+   */
+  public static parse(input: string): Version {
+    const result = this.decode(input)
+    if (isErr(result)) throw result.toError()
```
Yeah, I usually use a rule of three for things like this (only generalize if you repeat yourself three times), but anyway, done.

queerviolet commented on src/serde/index.ts:
```diff
@@ -0,0 +1,57 @@
+  = Serialize<T, N> & Deserialize<T, N>
+
+export interface Serialize<T=any, N=any> {
+  serialize: SerFn<T, N>
```
Yeah, it's unused right now, but we want to use this in the composer, and there, we will want to serialize nodes. I agree that it's troubling that there's a significant amount of untested code in the serialization path, but the intent is that we'll be testing and using it quite soon (and the tests can come before landing if you like!)

glasser replied:
"we want to use this in the composer" — right, not sure how I missed that :)

queerviolet commented on src/specs/core.ts:
```diff
@@ -0,0 +1,19 @@
+export const SpecUrl = scalar(core) ('SpecUrl', Spec)
+
+export default directive(core) ({
+  Using: repeatable({
```
So, that actually is the intent! I consider it a flaw that there's nothing forcing the spec definitions to match the binding files, and I expect that these spec binding files will soon be codegen'd from a machine-readable form of spec definitions ("feature schemas). I should probably make a note of this somewhere.

That did inform the design of this: while I wanted the eventually-codegen'd files to be relatively readable, it wasn't super important to me that people be able to deduce how to write them by hand. There were some earlier affordances to make things like `list()` and `must()` more discoverable, which I ditched specifically because they added complexity, and `Int.must.list` isn't much better than `must(list(Int)))`, except that autocomplete will suggest the former and not the latter (also, the latter arguably reads better).

I _do_ think it's quite nice that the types work out here in the interim, and that the conversion is checked. I don't _think_ it's possible to do that from e.g. `GraphQLDirective`s, since they turn the `ArgumentMap` into an array of `GraphQLArgument`s of unknown types, and their `toConfig` similarly returns a generic config, with the particular arguments and their types erased.

The other blocker / inconvenience with using higher-level / more graphql-jsey primitives like `GraphQLDirectives` to construct the binding is that I'm not sure how to represent everything we want to represent. For example, one thing you can do in a feature is extend metadata provided by other features. For example, the `join` feature can say:

```graphql
extend input core__error {
  join: join__Error
}
```

This lets you write `@core__error(join: { fieldsConflict: { ... } })`. This is nice because you can scan for all `core__error`s and find all the errors of any kind reported in a document (and, for example, quickly assert that there aren't any, or highlight them in explorer), but you can also provide structured error data from each spec. Here we happen to be extending a core directive, but there's no reason that has to be the case. Another example would be different transport features extending a `link__to` directive. A relevant consideration is that `@core__error` defines itself as being extensible-by-feature-name—i.e. features can add themselves as fields with their default name, and when the feature names are remapped, we'll also remap the field names. So if we're `@core(feature: ".../join/v1.0", as: "merge")`, then we would find `@core__error(merge: ...)` in the document. To get the `join`-specific data out of an error, we'd call e.g. `join.Error(core.Error(item))`, which handles the extraction.

I'm not totally sure how to represent that relationship with constructible `GraphQL*` objects, because e.g. extensions aren't even represented there. We could of course have a mixture of approaches, but that almost seems worse?

What I'd propose is: let me remove directive forms, which will simplify the binding layer quite a bit. Then let's evaluate from there?

martijnwalraven replied:
> Note that this could be built incrementally. We can first ship a version without codegen where it's your job to write the equivalent of the codegenned code yourself (which we could easily do for the small number of directives in core and join). Then while folks are making use of the library, we can build the codegen piece in parallel.

I agree with @glasser that these bindings add a lot of complexity to the codebase for currently not too much gain. Writing TypeScript definitions for the specs we care about by hand seems like a fine solution for now to be honest.

> I don't think it's possible to do that from e.g. GraphQLDirectives, since they turn the ArgumentMap into an array of GraphQLArguments of unknown types, and their toConfig similarly returns a generic config, with the particular arguments and their types erased.

While `GraphQLArgument` doesn't have generic arguments, so you can't derive type-safe code through TypeScript generic meta programming, it does [store its type](https://github.com/graphql/graphql-js/blob/995ea3a7ace7888502bee669c8dcecda520b0fa3/src/type/definition.d.ts#L567). So I may be missing something, but I don't really see the problem here if we would be using codegen? We could just parse SDL, and use the AST classes (`GraphQLDirective`, `GraphQLArgument`) to drive the generated code?

glasser replied:
Just a quick response to the "other blocker / inconvenience": the concern here is that you can't create the "schema"-level graphql-js classes from an `extend` without having the thing you would be extending?  My suggestion would be to base the core-schema-js APIs on "ast"-level graphql-js classes (like DirectiveDefinitionNode or InputObjectTypeExtensionNode), because they don't have that problem.  (So @martijnwalraven I think is being imprecise when he refers to GraphQLDirective/GraphQLArgument as "AST classes")

queerviolet commented on src/bind/directive.ts:
```diff
@@ -0,0 +1,235 @@
+ * { is: 'Using', Using: <using data> }
+ */
+export type Bind<F extends Forms> = {
+  [k in keyof F]: { is: k, on: ASTNode } & PickDeShape<F, k>
```
True. I regarded it as relatively minimal issue, since by convention the names of the feature input types are capitalized and `is` and `on` aren't. But I'll nest it when I remove directive forms.

queerviolet replied:
done

queerviolet commented on src/data.ts:
```diff
@@ -0,0 +1,95 @@
+  [getValue]: GetFn<T, S, Default>
+}
+
+export type Data<T, S=any, Default=undefined>
```
I'm open to making `Column` a class, but I do really like how both `data` and `derive` return typed accessor functions, which isn't really possible with a class. (Especially for `derive`, which is essentially just wrapper that memoizes a unary function.)

These things are used a lot more frequently than they're defined, so I optimized for friendly usage. `parentOf(node)` strikes me as quite clear. `parentOf.get(node)` seems weirder. `parent.get(node)` feels subtly confusing, and also awkward when you want to assign the result to the most obvious name (which is why we have `parentOf` and `documentOf` rather than `parent` and `document`).

I've added some extensive comments to `data` and `derive` that should hopefully make them feel less magical. I also stopped `export`ing `col`, since it's just a shared implementation detail between `data` and `derive`.

Do those changes address your general concern? I'm still happy to make `Col` a class if you still think it would be clearer, but it seems less important now.

queerviolet commented on src/bind/directive.ts:
```diff
@@ -0,0 +1,235 @@
+
+export default directive
+
+export function directive(spec: Spec) {
```
Nothing, i'll remove it when forms come out.

queerviolet commented on src/version.ts:
```diff
@@ -0,0 +1,140 @@
+    return `v${this.major}.${this.minor}`
+  }
+
+  private static VERSION_RE = /^v(\d+)\.(\d+)$/
```
Yeah, there's actually a test demonstrating that it accepts slightly out-of-spec version specifiers. Is that super bad? I can always add a check for it (probably in TS, since when I tried doing it in the regex it made the regex somewhat-stupidly more complicated).

glasser replied:
I realize Postel's law and all that, but if we're writing a spec and an implementation at the same time it seems reasonable to make them compatible.  I don't think people tend to write `v012` in practice (maybe v2021.04.01 ?) so being strict is unlikely to break folks?

No opinion about whether a check needs to be in re or TS

martijnwalraven commented on src/bind/directive.ts:
```diff
@@ -0,0 +1,235 @@
+  }
+}
+
+export type Layer<F extends Forms> = { spec: Spec }
```
I second this. I haven't given this quite as much attention as @glasser, but this is definitely one of the patterns that has made understanding this code difficult.

martijnwalraven commented on src/err.ts:
```diff
@@ -0,0 +1,228 @@
+    :
+    never
+
+export interface Ok<T> {
```
While I can see the usefulness of a `Result` type, I have trouble understanding these `Ok` and `Okay` types. What is the difference between them? Why do they have a `Maybe<ASTNode>` in addition to a `T`?

martijnwalraven replied:
It also seems there are plenty of existing `Result` implementations we could consider using. ([This package](https://github.com/supermacro/neverthrow) seems promising, but that's just one of the first results that came up so there might be better options.)

queerviolet replied:
`Ok` is the interface (exposed publicly). `Okay` is a class that implements the interface (not exposed). I tried just making `Ok` a class, but Typescript doesn't appear to treat the union of classes as a tagged union in some cases, which complicates the typings.

`Result` isn't a direct port of, say, Rust's `Result` enum. It's more fitted to this specific use case, where we're reporting the results of parsing or otherwise extracting something from a GraphQL AST.

In particular, I found that I was often mapping to a structure like `[node, result of parsing that node]`. The result of parsing the node shouldn't include the node (and sometimes can't, if for example it's a string). But it's useful to have it on the result (to filter or do other operations on). This seemed like a pretty straightforward way of achieving that.

martijnwalraven replied:
I'm not sure I understand this. Can you give a more detailed example of how this would be used? Why does that require putting `Maybe<ASTNode>` on `Ok` in addition to `T`?

queerviolet commented:
PTAL. I've removed directive forms, which simplifies the directive logic a lot. I've also moved the scanning logic into its [own module](./src/scan.ts). The result of `scan`ing a node for a directive is now an array of `SpecifiedData`, which is a type tagged with the spec URL and directive name, distinguishing different kinds of metadata in a clear and type-friendly way.

I think the separation will make it easier to specify these things using `GraphQLDirective`, `GraphQLInput`, etc, though tbh I'm still struggling to see how that'll work or why we want to do it. Anyway, take a look at what's there, and I'll take another crack at changing the definition layer to `GraphQLDirective`, `GraphQLScalar`, `GraphQLInput`, etc. Though honestly, if someone who understands the objection better wanted to take a crack at that, that would be fine too.

queerviolet commented on src/err.ts:
```diff
@@ -0,0 +1,228 @@
+import { documentOf, sourceOf } from './linkage'
+
+/**
+ * An Err is an error on a GraphQL document.
```
Oh, I actually didn't know `graphql-js` had a source mapping utility. We can definitely switch to using that!

martijnwalraven replied:
Would that also include switching to `GraphQLError` instead of a custom `Err` type?

martijnwalraven commented:
> I think the separation will make it easier to specify these things using `GraphQLDirective`, `GraphQLInput`, etc, though tbh I'm still struggling to see how that'll work or why we want to do it. Anyway, take a look at what's there, and I'll take another crack at changing the definition layer to `GraphQLDirective`, `GraphQLScalar`, `GraphQLInput`, etc. Though honestly, if someone who understands the objection better wanted to take a crack at that, that would be fine too.

I'd love to find a way to help here, but to be honest I still don't fully understand what the current code is doing (and why it is doing that).

My understanding is that the main goal of this library is to extract spec-specific data from parsed `ASTNode`s. That involves finding relevant `DirectiveNode`s (based on the local name of the spec in the current document, as specified using `@core`) and extracting their argument values.

What I'm unsure about is why that requires SerDe and the rest of the binding machinery. This may be naive, but couldn't we just rely on `GraphQLDirective` (either constructed through code or parsed from SDL) and use [`graphql-js`'s `getArgumentValues()`](https://github.com/graphql/graphql-js/blob/735975d4904c5d9cc93d0c0a02af181e8f02dfba/src/execution/values.js#L149-L159) to extract argument values (which takes care of type checking and makes sure required values are provided)? 

We could then write manual TypeScript definitions for the extracted values, which honestly seems like a fine solution (this basically reiterates the point @glasser makes in https://github.com/apollographql/core-schema-js/pull/2#discussion_r575740887).

A missing piece here might be local namespace support, but it seems we could get that by modifying names in `GraphQLDirective`s based on the spec definitions in the document. Or maybe we need a modified version of `getArgumentValues()` that is aware of name mappings.

I could very well be simplifying this too much, but if that is the case it would be really helpful to get a better understanding of the additional use cases we're trying to support with this.

queerviolet commented:
Ah, see, I didn't know about `getArgumentValues()`. "You should use `getArgumentValues()` here" is actionable feedback that I can implement! Similarly for the `Source` stuff.

(I went looking for both of these, but I find `graphql-js`'s documentation… pretty hard to navigate)

martijnwalraven commented:
> Ah, see, I didn't know about `getArgumentValues()`. "You should use `getArgumentValues()` here" is actionable feedback that I can implement! Similarly for the `Source` stuff.

Good to hear! I just noticed `getArgumentValues()` is marked `@internal`, which doesn't stop you from using it but would require a deep import. The related `getDirectiveValues()` does seem to be part of the public API, and it calls into `getArgumentValues()`. If neither works, you could also copy/adapt the code from `getArgumentValues()`. What is important here I think is that we can treat `GraphQLDirective` as the source of truth for the directive definition, including the argument typing information.

> (I went looking for both of these, but I find `graphql-js`'s documentation… pretty hard to navigate)

Yeah, there is no comprehensive documentation to speak of unfortunately. The stuff at https://graphql.org/graphql-js/ is incomplete and out of date. But the source is pretty well organized, with sensibly named functions and inline documentation. It may require some digging, but I find `graphql-js` generally has the functions I need or at least shows me a way of approaching something.

glasser commented on src/namespace.ts:
```diff
@@ -0,0 +1,114 @@
+
+export function namespaceFor(doc: DocumentNode, spec: Spec): Maybe<Namespace> {
+  for (const ns of namespacesIn(doc).values()) {
+    if (ns.spec && spec.satisfies(ns.spec))
```
Do we have to be worried about the case where multiple namespaces could be satisfied by the same spec (ie, multiple `@core` with the same identity and major version but different minor versions)?  Or do we block that elsewhere?

glasser reviewed (commented):
Sorry, I didn't make it all the way through this round of review before I need to head out for some holiday celebrations. Might get back to it later tonight or definitely tomorrow!

glasser commented on src/data.ts:
```diff
@@ -0,0 +1,167 @@
+ *   const someObject = {}
+ *   set(someObject, theMessage, 'hello world')
+ *   theMessage(someObject)      // -> 'hello world'
+ *   // or with get():
```
BTW just to be explicit, my concern isn't "function types are more clear than a generic type that basically represents a function": nothing wrong with `Read<V, S>`. It's "because you can invoke it in two different ways, the type tells you both of those ways every time".

One other note: as far I can tell the *only* use you make of the Read default value is internally in the implementation of `derived`. So this ends up being an extra distracting type parameter to Read, ReadFn, etc, even though it's completely unused for `data`. This is another thing that makes me feel like just having `data` and `derived` have their own implementations that don't try to share a common helper would make things simpler.

queerviolet replied:
Ok, I think I'll just take your suggestion of removing `data` entirely, which both simplifies things and frees up the name `Data` to use instead of `SpecifiedData`.

glasser commented on src/scan.ts:
```diff
@@ -0,0 +1,99 @@
+          return []
+        }
+        const [errors, values] = siftValues(results)
+        report(...errors)        
```
I don't really understand error handling in this function.  If the error is "multiple directives on non-repeatable" then we stash that error on the doc and return empty list to the caller.  But if the error is "we couldn't parse something" then we stash the error on the doc and we still return data to the caller.  And then this function only has one caller (in isExport) which just looks at the return value without doing anything.  It seems weird to me that that caller should be just carrying on with its business even if what it saw was `core__surface(export: 123)`? I guess eventually the user is responsible for calling `ensure` but it feels like failing faster makes sense here?

glasser commented on src/schema.ts:
```diff
@@ -0,0 +1,197 @@
+/**
+ * Attach metadata layers to a document.
+ * 
+ * Calling this isn't required, but ensures that the provided layers are
```
I only sort of get why this isn't required. I guess the point is that you can just come along whenever and apply a feature to the document and see what happens, but if you do it all at once at the front then the recommended error handling pattern works.

It's not that "calling this" is what ensures the provided layers are scanned" — it's the pattern of attaching the layers at all (either manually or via this function) *before calling ensure() and otherwise using the layers* that ensures it?

I frankly think it would be easier to understand if we had an API that was just "here's a Source and some layers, apply them and handle errors" instead of breaking it into pieces like this which the user has to be sure to apply in the right order, but maybe this isn't the most high leverage place to be asking for changes at this stage

glasser commented on src/schema.ts:
```diff
@@ -0,0 +1,197 @@
+ * 
+ * @param asSource 
+ */
+export function fromSource(...asSource: AsSource): Pipe<DocumentNode> {
```
Even though it's now documented I still don't think it's helpful for the main entry point to require learning about Pipe

glasser commented on src/schema.ts:
```diff
@@ -0,0 +1,197 @@
+ * Errors in this document
+ */
+export const errors = derive('Document errors',
+  (_: DocumentNode): Err[] => [])
```
You know, if you're fine with the pattern of using `derive` to attach an empty mutable object to a node, then you can 100% delete `data` and replace `derive`.  Here, I gave it a shot at https://github.com/apollographql/core-schema-js/commit/1be287dc80be55e6e02c35e2ceb8886517c5f979

I think data.ts in that version is way easier to understand. (Probably rename to derive.ts.)

glasser commented on src/schema.ts:
```diff
@@ -0,0 +1,190 @@
+            schema = def
+            continue
+          }
+          const error = ErrExtraSchema({ doc, node: def })
```
This hasn't been done

glasser commented on src/schema.ts:
```diff
@@ -0,0 +1,190 @@
+
+    const {ok: coreUse, node: directive} = coreReq
+
+    if (coreUse.using.identity !== core.identity) {
```
How about this?

glasser commented on src/schema.ts:
```diff
@@ -0,0 +1,190 @@
+        e => e.node?.kind === 'Directive' &&
+        e.node.name.value === coreName)
+    )
+    return okays.map(r => r.ok)
```
Same here... maybe my schema.ts notes in general weren't seen?

glasser commented on src/source.ts:
```diff
@@ -0,0 +1,59 @@
+  function describe(from?: Location | ASTNode) {
+    const loc = locationFrom(from)
+    if (!loc) return `${src}:[unknown]`
+    const line = search.closest(endings, loc.start)
```
I don't think any of my comments on this file were noted (including this one which returns bogus output), though maybe we're switching to GraphQLError / locations?

glasser commented on src/spec.ts:
```diff
@@ -0,0 +1,77 @@
+    const url = result.ok
+    const path = url.pathname.split('/')
+    const verStr = path.pop()
+    if (!verStr) return ErrNoVersion({ url })
```
Not sure my comments on spec.ts were seen

glasser commented on src/serde/index.ts:
```diff
@@ -0,0 +1,57 @@
+ * These should be declared with `Serialize` and `Deserialize` directly, rather than
+ * as `SerDe`.
+ */
+export type SerDe<T=any, N=any>
```
Still curious if we can make N be ValueNode and ValueNode|DirectiveNode based on my concerns about how we end up constructing stuff in recursive serializers.

glasser commented on src/bind/directive.ts:
```diff
@@ -0,0 +1,235 @@
+        const requests = using(doc)
+        for (const r of requests) {
+          if (r.using.identity === spec.identity &&
+            spec.version.satisfies(r.using.version)) {
```
You've refactored the API around directives a lot but I'm still not sure how you would write code that says "match this doc for spec with a certain identity and a version that's either 0.1 or 0.2 or 0.3".

glasser commented on src/bind/index.ts:
```diff
@@ -0,0 +1,9 @@
+
+export interface Specified {
+  spec: Spec
+  name: string
```
OK I thiiiink based on this being more used that this is the specified name with the original prefix? I guess that's kinda implied by the interface being called Specified but still could use a comment

glasser reviewed (commented):
OK I think I've finished a pass here.

glasser commented on src/spec.ts:
```diff
@@ -0,0 +1,90 @@
+    const verStr = path.pop()
+    if (!verStr) return ErrNoVersion({ url })
+    const version = Version.parse(verStr)
+    const name = path[path.length - 1]
```
Should we validate that name is a valid prefix (eg, it contains GraphQL-identifier-friendly characters?)

glasser commented on src/namespace.ts:
```diff
@@ -0,0 +1,114 @@
+  'Namespaces in document', (doc: DocumentNode) => {
+    const names = new Map<string, Namespace>()
+    for (const layer of using(doc)) {
+      const name = layer.as ?? layer.using.name
```
Is this an appropriate place to check that `as` arguments are valid GraphQL identifier characters?

abernix commented:
Administrative note: This PR is "merged" because the PR that superseded it (#3) was finished, approved and merged in as a true Git merge (not a squash).  Since this PRs ab6c7b6 was included in that PR / its root, this PR is effectively merged even though there was heaps of follow up.

(Fwiw, I might suggest a rebase or a squash in the future given how much of this PR was reverted, but not asking for any change.)

queerviolet commented:
damnit i really thought i'd squashed it

On Mon, Aug 2, 2021 at 8:05 AM Jesse Rosenberger ***@***.***>
wrote:

> Administrative note: This PR is "merged" because the PR that superseded it
> (#3 <https://github.com/apollographql/core-schema-js/pull/3>) was
> finished, approved and merged in as a true Git merge (not a squash). Since
> this PRs ab6c7b6
> <https://github.com/apollographql/core-schema-js/commit/ab6c7b6d2c84cd9fe5a774f64bf662f508310d18>
> was included in that PR / its root, this PR is effectively merged even
> though there was heaps of follow up.
>
> (Fwiw, I might suggest a rebase or a squash in the future given how much
> of this PR was reverted, but not asking for any change.)
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/apollographql/core-schema-js/pull/2#issuecomment-890970433>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB3HLTVCAG5UYJ2TCRDHKE3T22CXHANCNFSM4XLY2VWQ>
> .
>

abernix commented:
@queerviolet All good!  I could conceivably still "fix" that if I move this into the `federation` repository.)
